src/mod_stream_management.erl

```diff
@@ -128,17 +293,74 @@ register_smid(SMID, SID) ->
               {error, Reason}
     end.
 
-get_sid(SMID) ->
-    case mnesia:dirty_read(sm_session, SMID) of
-        [#sm_session{sid = SID}] ->
-            [SID];
-        [] ->
-            []
+register_stale_smid_h(SMID, H) ->
+    case gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, {false, []}) of
+        {false, []} -> ok;
+        {true, _} ->
+            try
+                Stamp = now_to_seconds(erlang:timestamp()),
+                mnesia:sync_dirty(fun mnesia:write/1,
+                                  [#stream_mgmt_stale_h{
+                                      smid = SMID, h = H, stamp = Stamp}])
+            catch exit:Reason ->
+                      {error, Reason}
+            end
+    end.
+
+remove_stale_smid_h(SMID) ->
+    case gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, {false, []}) of
+        {false, []} -> ok;
+        {true, _} -> mnesia:dirty_delete(stream_mgmt_stale_h, SMID)
     end.
 
+%%
+%% gen_server
```

Comment:
Two things regarding the gen_server:
1. I think it'd be better to move it to a separate module.
2. In cluster setup we'll have such a process on every node in the cluster. Did you check what happens if we run the `select_delete` in the same time? I know the chances are rather low.